Refactored build-script-impl for cross-compiling support

karwa opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->
In the current build-script-impl, we use the variable name "deployment_target" pretty loosely when talking about the Build, Host and Target of the products we're building. You can see this in places - like where the test phase iterates every STDLIB_DEPLOYMENT_TARGET (i.e. Target), before skipping those which are not, effectively, host platforms, so we're tripping up on all these types of "deployment targets" floating around.

I'm aware the build script is being rewritten in python. The rewrite is approaching the stage where it's going to be getting to some of the points I've addressed here (e.g. calculating stdlib targets), so I think it's a good idea if we get this done first, so the python implementation can consider it from the start.

This PR replaces uses of "deployment_target" with "host" when talking about a platform which will ultimately host a Swift compiler. The Build platform is implicitly a Host. It replaces uses of "deployment_target" with "stdlib_target" when talking about a platform for which the stdlib is to be built. This terminology is consistent with the actual swift/llvm build process: `llvm_host_triple`, `SWIFT_HOST_VARIANT_*` and host-related flags get set `set_build_options_for_host` instead of `set_deployment_target_options`, and `SWIFT_STDLIB_TARGETS` calculation has been moved to a function - `calculate_targets_for_host`.

- `SWIFT_STDLIB_TARGETS` now gets recalculated for each host we build, as it should.
- Executables are always searched for in the `LOCAL_HOST` (Build machine) products
- Some switching on 'uname' with switches on the host, in support of future cross-compiling
- Removed lipo step (it lipos stuff between hosts, which doesn't make sense - possibly a result of confused terminology again?)

_Note_: The diff is large because of things like moving the stdlib target calculation in to a function and relocating it to be close to the function which sets the host options. The actual changes are not that extreme, and really limited to some variable renaming, function encapsulation and moving.

<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

- Use consistent terminology for target platforms, disambiguate uses of
  'deployment_target'
- Calculate targets per host, in preparation for cross-compiling
- Only use executables from LOCAL_HOST
- Only execute testable targets of the LOCAL_HOST
- Remove some cases of switching on 'uname' in preparation for
  cross-compiling

karwa commented:
Diff is now more manageable

gribozavr commented on utils/build-script-impl:
```diff
@@ -316,55 +316,73 @@ function num_swift_parallel_lto_link_jobs() {
             ;;
+        linux-*)
+        SWIFT_HOST_VARIANT="linux"
+        SWIFT_HOST_VARIANT_SDK="LINUX"
```
Missing indentation?

gribozavr commented on utils/build-script-impl:
```diff
@@ -1475,8 +1554,28 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
+        if [[ $(is_cross_tools_host ${host}) ]] ; then
+
+            # Cross-compiled hosts are always compiled after the LOCAL_HOST,
+            # and are built with the freshly-baked tools we just made.
```
Why is it a requirement to build the cross-compiler with the just-built tools?  This does not work for our internal workflows.

karwa replied:
There was some talk about cross-compiling before and I believe that was what was suggested. It doesn't have to be; but it's probably better as the default. I could add a flag to bypass it.

gribozavr replied:
We would need to build just the runtime library with the just-built compiler.  I'd rather limit the amount of stuff we are building with the just-built clang.

gribozavr commented on utils/build-script-impl:
```diff
@@ -565,21 +573,32 @@ function set_deployment_target_based_options() {
+    fi
+    swift_cmake_options+=(
+        -DSWIFT_HOST_VARIANT_SDK="${SWIFT_HOST_VARIANT_SDK}"
+        -DSWIFT_HOST_VARIANT_ARCH="${SWIFT_HOST_VARIANT_ARCH}"
```
I'm afraid CMake files are not prepared to accept `SWIFT_HOST_VARIANT_ARCH` from the command line and would just unconditionally overwrite it.

karwa replied:
You're right, I've got another patch which cleans up that part of the CMake files. It uses `SWIFT_HOST_VARIANT_SDK` and `SWIFT_HOST_VARIANT_ARCH` to calculate the Swift compiler host/stdlib targets (instead of directly parsing the `CMAKE_SYSTEM_*` variables like we do today), but falls back to parsing from the CMAKE_SYSTEM if they're not set.

I could add it to this PR or do it separately; It's self-contained.

gribozavr replied:
It seems to me that it would be best if we could land the CMake changes ahead of this PR -- is that possible?

karwa replied:
https://github.com/apple/swift/pull/2560

gribozavr commented on utils/build-script-impl:
```diff
@@ -565,21 +573,32 @@ function set_deployment_target_based_options() {
+    fi
+    if [[ "${SWIFT_HOST_VARIANT}" ]] ; then
+        swift_cmake_options+=(
+            -DSWIFT_HOST_VARIANT="${SWIFT_HOST_VARIANT}"
```
Please add `:STRING`.

gribozavr commented:
The primary reason why I didn't use the Build/Host/Target terms is because it assumes one combination of the three, and does not work well with multi-target cross-compilers compilers, which Swift is.

gribozavr commented on utils/build-script-impl:
```diff
@@ -2383,41 +2490,16 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
-done
-
-if [[ "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARGETS}" ]] && [[ ! "${SKIP_MERGE_LIPO_CROSS_COMPILE_TOOLS}" ]] ; then
-    echo "--- Merging and running lipo for ${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARGETS[@]} ---"
```
Did you intentionally drop this part?  It is important for our internal workflows.  Is this handled somehow by the primary build steps now?

karwa replied:
Yes, but looking at it more closely, I should have put it behind a switch for the iOS targets.

gribozavr replied:
The purpose of this step is to lipo together the cross-compiled compilers, and copy over the standard library, so that we would get a consistent installation layout ready to be deployed onto a target.  This step is Darwin-specific, but other targets will need something similar.

karwa replied:
I've added the lipo step back. The install process needed some adjustment to deal with it. Since it's a collection of multiple hosts, it gets placed in to a host-specific subdirectory "merged".

gribozavr commented on utils/build-script-impl:
```diff
@@ -2210,6 +2304,13 @@ for deployment_target in "${STDLIB_DEPLOYMENT_TARGETS[@]}"; do
 
+        # We can only run tests built for the host machine, because
+        # cross-compiled hosts only build their native target. See: get_stdlib_targets_for_host()
+        if [[$(is_cross_tools_host ${host})]]; then
```
Please add spaces inside of `[[ ]]`.

gribozavr commented on utils/build-script-impl:
```diff
@@ -2012,28 +2116,18 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
-            exit 1
-            ;;
-    esac
+for host in "${ALL_HOSTS[@]}"; do
```
Why did you change this to `ALL_HOSTS`?  When compiling on OS X, we are also cross-compiling for iOS simulator, for example, and we want to run tests for both OS X and iOS simulator; that is, for every stdlib target.

karwa replied:
Aren't the simulator test targets build products of the OSX host? they are in, e.g. `build/Ninja-XXX/swift-macosx-x86_64/test-iphonesimulator-i386`

We want to iterate over every host we built products for, and call `calculate_targets_for_host` to set SWIFT_TEST_TARGETS, and then do the tests with them.

Iterating over `STDLIB_DEPLOYMENT_TARGETS` is incorrect here - those targets will change for every host. Doing this allows us to cross-compile the unit tests which statically link against the runtime.

gribozavr replied:
I see.  I think I understand now.

gribozavr commented on utils/build-script-impl:
```diff
@@ -1001,138 +997,216 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+    done
+
+    # Filter duplicate SWIFT_SDKs (which will exist if building for multiple architectures)
+    SWIFT_SDKS=(`echo "${SWIFT_SDKS[@]}" | tr " " "\n" | sort -u | tr "\n" " "`)
```
`SWIFT_SDKS` is a command line parameter for `build-script-impl` that is used to limit the SDKs that we would otherwise automatically discover.  For example, on OS X, we would automatically build for iOS, iOS simulator, watchOS, watchOS simulator etc.  You can pass `--swift-sdks=OSX` to limit the SDKs we configure to just OS X.

The reason why this is needed is that we build Swift in environments where a certain SDK (that we don't need to build for) appears to be available and would be detected and configured, but is actually unusable and would fail the build.

karwa replied:
Yes, but this is also what we'd want to communicate when cross-compiling. For example, I can tell it to build the `LINUX` (i.e. host's native) and `ANDROID` SDKs, or maybe some other targets.

We would need to rework how we specify SDKs to support more target SDKs (e.g Android/x86), but this is the place it would go.

gribozavr replied:
Reusing the variable makes sense to me, but changing the format would be painful -- we have existing build presets that pass semicolon-separated lists here.  Could you revert to accepting semicolon-separated lists, or either kind of lists?

gribozavr replied:
Also, it looks like the line `SWIFT_SDKS=()` above would overwrite the command line parameter, wouldn't it?

karwa replied:
I didn't notice there was a parameter for that, but yes it would. Is it necessary to specify the SDKs to build, given that they're all derived from the stdlib targets, which is already a parameter?

rintaro replied:
`--swift-sdks` is to "configure", `--stdlib-deployment-targets` is to "build"
If we don't specify `--swift-sdks`, it should be automatically discovered as all supported SDKs.
Although I'm personally not sure why we need to limit "configure"ed SDKs,
but possibly, we may use this to *expand* it.
```
build-script --swift-sdks="LINUX;EDISON" \
  --custom-cmake-options="-DSWIFT_SDK_EDISON_PATH=/path/to/edison_sdk" \
  --custom-cmake-options="-DSWIFT_SDK_EDISON_LIB_SUBDIR=edison" \
  ...

gribozavr replied:
@rintaro Exactly, `--swift-sdks` is for configuration.

As for why,

> The reason why this option is needed is that we build Swift in environments where a certain SDK (that we don't need to build for) appears to be available and would be detected and configured, but is actually unusable and would fail the build.

gribozavr commented:
After reading the script more, I find the change to the plain "host/target" terms to be a readability loss.  In the current script, I tried to use "stdlib-deployment-target" and "tools-deployment-target" to be extra clear about what is happening.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
Some strange CI issues.  Let's try again.

@swift-ci Please test

gribozavr commented:
@karwa I just wanted to thank you and say that I deeply appreciate your work on the build system.

karwa commented:
Happy to help!

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

shahmishal commented on utils/build-script-impl:
```diff
@@ -316,55 +316,74 @@ function num_swift_parallel_lto_link_jobs() {
+                    ;;
+                linux-armv7)
+                    SWIFT_HOST_VARIANT_ARCH="armv7"
+                    llvm_host_triple="${SWIFT_HOST_VARIANT_ARCH}-linux-gnueabihf"
```
Please substitute `SWIFT_HOST_VARIANT_ARCH` with `armv7` for clarity.

karwa replied:
I wanted to emphasise that the LLVM and Swift triples should be exactly the same. I've had issues with an "armv7-linux" Swift triple and "arm-linux" LLVM triple, for example. There is probably a better way to do this, such as a single "triple" setting that gets passed to both LLVM and Swift (we can now set `SWIFT_HOST_TRIPLE` explicitly when building, so that's an option)

shahmishal commented on utils/build-script-impl:
```diff
@@ -565,21 +574,32 @@ function set_deployment_target_based_options() {
+            -DLLVM_HOST_TRIPLE:STRING="${llvm_host_triple}"
+        )
+    fi
+    if [[ "${SWIFT_HOST_VARIANT}" ]] ; then
```
When is `SWIFT_HOST_VARIANT` not set? In the switch statement above its set for all cases (all platforms).

shahmishal commented on utils/build-script-impl:
```diff
@@ -1001,138 +998,245 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
-        fi
+}
+
+# When building cross-compilers for these hosts, merge all of their contents together with lipo
```
80 columns please.

shahmishal commented on utils/build-script-impl:
```diff
@@ -1001,138 +998,245 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+    SKIP_COMPILER_RT=1
+fi
+
+STDLIB_DEPLOYMENT_TARGETS_DEFAULT="${STDLIB_DEPLOYMENT_TARGETS}"
```
Where does `STDLIB_DEPLOYMENT_TARGETS` get assigned? Also, why do we need to copy it to `STDLIB_DEPLOYMENT_TARGETS_DEFAULT`?

karwa replied:
`STDLIB_DEPLOYMENT_TARGETS` is a parameter to the build script. You're right that the copy in to `_DEFAULT` is useless; all iteration of the stdlib deployment targets is through this function (`get_stdlib_targets_for_host`)

shahmishal commented on utils/build-script-impl:
```diff
@@ -2265,17 +2440,17 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
                         set_lldb_build_mode
                         pushd ${LLDB_SOURCE_DIR}
-                        xcodebuild -target toolchain -configuration ${LLDB_BUILD_MODE} install ${lldb_xcodebuild_options[@]} DSTROOT="${INSTALL_DESTDIR}" LLDB_TOOLCHAIN_PREFIX="${TOOLCHAIN_PREFIX}"
+                        xcodebuild -target toolchain -configuration ${LLDB_BUILD_MODE} install ${lldb_xcodebuild_options[@]} DSTROOT="${host_install_destdir}"
```
@scallanan Could you review this change?

The background is that `host_install_destdir` is now equal to `${INSTALL_DESTDIR}/${TOOLCHAIN_PREFIX}`.  Does removing `LLDB_TOOLCHAIN_PREFIX` cause any issue while building toolchain?  I don't see any issue if LLDB only needs to install files into the toolchain directory.  I verified that in the toolchains that we upload to swift.org there are no files outside of the toolchain directory.

karwa replied:
When I tried it with the `LLDB_TOOLCHAIN_PREFIX` parameter, I got a strange output tree. In order to build a clean toolchain in both situations (cross-compiling and OSX-only), I had to:
- Omit the `LLDB_TOOLCHAIN_PREFIX` parameter. It seems to not always append a path separator, meaning I got output files in '${INSTALL_DESTDIR}${TOOLCHAIN_PREFIX}" - i.e "destdirswift3.xctoolchain"
- Include the `TOOLCHAIN_PREFIX` in the install location. Otherwise the swift compiler itself would install itself in to a `usr` folder adjacent to the toolchain location

I can't explain how it produced correct toolchains before. Both changes make sense to me. [This is a dump of the strange tree](https://github.com/apple/swift/files/281855/strange_tree.txt) I was getting without them.

.

gribozavr replied:
> It seems to not always append a path separator

This part can be fixed.

> In order to build a clean toolchain in both situations (cross-compiling and OSX-only), I had to:

What were the build and host machines?

The problem here is that the compiler wants to install itself into `<prefix>/bin/swift`, but we want it in `<toolchain>/usr/bin/swift`, so we set `prefix=<toolchain>/usr`.  But LLDB needs to install files into `<toolchain>/System`, so it has to know the path to the toolchain directory.  LLDB's Xcode build system also has native knowledge of toolchains, so I'd be very hesitant to avoid using `LLDB_TOOLCHAIN_PREFIX`.

karwa replied:
Build and host both OSX. Before the changes, we install the compiler with the command:
`DESTDIR="${target_install_destdir}" "${CMAKE}" --build "${build_dir}" -- ${INSTALL_TARGETS}`
where `target_install_destdir` is just `INSTALL_DESTDIR` if you're not cross-compiling (e.g. OSX on OSX). I don't see where the swift compiler install command even gets told about the toolchain prefix.

gribozavr replied:
> I don't see where the swift compiler install command even gets told about the toolchain prefix.

I believe it is this line:

```
-DCMAKE_INSTALL_PREFIX:PATH="${INSTALL_PREFIX}"
```

(line 1582 on the master branch)

karwa replied:
That means `INSTALL_PREFIX` has to include `TOOLCHAIN_PREFIX`, which isn't documented anywhere. Also, it's problematic because we would then need host-specific install prefixes when cross-compiling (which is probably a good idea, if a little awkward to implement. We can avoid it for now if all hosts use `/usr` and we insert the toolchain prefix in the build script)

gribozavr replied:
> That means `INSTALL_PREFIX` has to include `TOOLCHAIN_PREFIX`, which isn't documented anywhere.

It is true that it is not documented, but that's how existing presets work.  (If we agree to change behavior we would need to update all of them.)

> Also, it's problematic because we would then need host-specific install prefixes when cross-compiling (which is probably a good idea

Sure.

> We can avoid it for now if all hosts use /usr and we insert the toolchain prefix in the build script

As long as existing presets continue to work, I'd be OK with it; although I remember someone has posted toolchains for Linux that install into `/usr/local`, so eventually we'd like to get this configurable again.

gribozavr replied:
A possible course of action is to add an option called `darwin-toolchain-path` that would be set to `/Applications/Xcode.app/.../FooBar.xctoolchain` when building a compiler for OS X and unset everywhere else.

karwa replied:
There would only be a problem if you're on OSX, building a compiler for a linux host, and you wanted that linux host to have some other install prefix than `/usr`. Then the OSX toolchain that gets placed in the install location wouldn't have the correct prefix.

For non-cross-compilers, the changes would be for OSX: you would always specify a `/usr` install prefix, strip the toolchain prefix from any scripts. The products will still be in the same location as before, just the parameter changes.

Alternatively, we could hard-code the `/usr` install prefix for Darwin toolchains and ignore the `INSTALL_PREFIX` parameter.

karwa replied:
LLDB shouldn't be a problem, either - we're setting the `DSTROOT` parameter when calling `xcodebuild`, which is where the products will be placed when installing. Not passing the prefix parameter shouldn't be a problem.

scallanan replied:
`LLDB_TOOLCHAIN_PREFIX` is part of the Xcode-based build system that ensures both that internal Apple builds work and also that open-source builds work correctly.

I'm not sure about all the moving parts in this patch, but that variable definitely should be passed to `xcodebuild`.

karwa replied:
Okay, it's back. Adding a '/' before the toolchain prefix was enough to fix the install issue.
Also removed toolchain_prefix from the install_prefix in the OSX build preset.

shahmishal commented on utils/build-script-impl:
```diff
@@ -2328,8 +2496,8 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
-                XCTEST_BUILD_DIR=$(build_directory ${deployment_target} xctest)
-                XCTEST_INSTALL_PREFIX="${INSTALL_DESTDIR}/${INSTALL_PREFIX}/lib/swift/${LIB_TARGET}"
+                XCTEST_BUILD_DIR=$(build_directory ${host} xctest)
+                XCTEST_INSTALL_PREFIX="${host_install_destdir}/${INSTALL_PREFIX}/lib/swift/${LIB_TARGET}"
```
Sorry, I'm confused by this change.  On OS X, `host_install_destdir` is the full path to the toolchain directory on the build machine.  We then append `INSTALL_PREFIX` to it, which is the full path to the toolchain.  What was your intent behind this change?

I think we just want to append the `usr` path component, but we shouldn't hardcode that.

I think we can just keep the code the old way, what do you think?

gribozavr replied:
@karwa Sorry, did you see this comment?

karwa replied:
on OSX, `host_install_destdir` will be the path to the toolchain root, and `INSTALL_PREFIX` will always be `/usr`, not the full path to the toolchain.

This should resolve to `/destdir/toolchain-prefix/usr/lib/swift/macosx`, which is correct

shahmishal commented on utils/build-script-impl:
```diff
@@ -1193,6 +1297,46 @@ function build_directory_bin() {
     fi
 }
 
+function get_host_install_destdir() {
```
Could you make this function echo the value instead of setting a global?  It would be more in line with other functions in this file that are callable via `result=$(foobar)` syntax.

shahmishal commented on utils/build-script-impl:
```diff
@@ -1001,138 +998,245 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+function calculate_targets_for_host() {
+
+    SWIFT_STDLIB_TARGETS=()
+    SWIFT_SDKS=()
```
Sorry, I'm afraid this issue with `SWIFT_SDKS` is not resolved yet.  `build-script` supports two workflows:

1. Configure all SDKs, build on demand.  On OS X, developers typically run `./build-script -R` to just build the OS X standard library, and then (some time later) run `./build-script -R -i -t` to build the iOS library and run tests.  What's important is that the first invocation (`./build-script -R`) will discover and configure all SDKs (OS X, iOS, tvOS, watchOS), but will only build for OS X.  Then, running the second command will not cause CMake to be re-run.

2. Configure the exact list of SDKs (using the `--swift-sdks` option), and build on demand.  This is similar to (1), but the developer chooses to not even configure one of the SDKs (it is available, but is broken and will fail the build if you try to use it).

karwa replied:
How about instead of `swift-sdks` as a parameter to `build-script-impl`, we rename it to `build-stdlib-deployment-targets`. It would be more descriptive, and would specify the same kind of details as the existing `stdlib-deployment-targets` parameter. That would also allow us to only build specific slices (e.g. only `iphone-armv7`, instead of all `iphone-*` architectures), which I believe is something the core team is interested in.

gribozavr replied:
> How about instead of `swift-sdks` as a parameter to `build-script-impl`, we rename it to `build-stdlib-deployment-targets`.

Sorry, could you clarify what you are suggesting?  We need to keep the `swift-sdks` with the old semantics (limit the SDKs to configure), we can rename it though.

Are you suggesting that we would add another parameter that would control which SDKs would be built, a subset of those which are configured?  This sounds good to me.

> That would also allow us to only build specific slices (e.g. only `iphone-armv7`, instead of all `iphone-*` architectures), which I believe is something the core team is interested in.

There's unfortunately some complexity here.  We can't dynamically change the list of slices after configuration, because rules generated by CMake hardcode the list of standard library slices to be lipo'ed together.  This would have to be a configuration parameter.

karwa replied:
As I've implemented it now, `stdlib-deployment-targets` controls which targets get configured and `build-stdlib-deployment-targets` controls which get built (default is `all`, meaning all configured, although this is in conjunction with the various `skip-ios`..etc flags as we have them today -- it's an additional filter).

So when it comes to rebuild, it'll invoke the following command (for OSX and iOS-armv7):
```
+ /usr/local/bin/cmake --build /Spring/Projects/3rdParty/OpenSource/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64 -- -j8 all swift-test-stdlib-macosx-x86_64 swift-test-stdlib-iphoneos-armv7 swift-benchmark-macosx-x86_64 swift-benchmark-iphoneos-armv7
```
Unfortunately, the target `swift-test-stdlib-iphoneos-armv7` depends on universal `iphoneos` libraries rather than its own architecture-specific variants, so it ends up building all architecture slices for the SDK.

shahmishal commented:
@karwa Can you please fix the conflicts?

rintaro commented on utils/build-script-impl:
```diff
@@ -1001,138 +998,245 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+    # Filter duplicate SWIFT_SDKs (which will exist if building for multiple architectures)
+    SWIFT_SDKS=(`echo "${SWIFT_SDKS[@]}" | tr " " "\n" | sort -u | tr "\n" " "`)
+
+    echo "Building the standard library for: ${SWIFT_STDLIB_TARGETS[@]}"
```
This `calculate_targets_for_host ` is called two times for each `host`, "configure/build" and "test".
I don't feel good about printing these messages here.

karwa commented:
OK I've tested that we can build on OSX and it configures for all available SDKs but only builds OSX. You can turn on iOS with the `--ios` flag and it will build those too, without having to reconfigure. You can also limit the slices to build with `build-stdlib-deployment-targets`... in theory. The ninja command is limited appropriately, but our CMake configuration files are not yet that granular. So limiting the build to `macosx-x86_64;iphoneos-armv7` will build for OSX and all iphoneos slices.

We would need to change our configuration to be more granular if we wanted to limit building any further:

```
build stdlib/public/core/swiftCore-iphoneos: 
    phony stdlib/public/core/CMakeFiles/swiftCore-iphoneos
    lib/swift/iphoneos/libswiftCore.dylib
    lib/swift/iphoneos/arm64/libswiftCore.dylib
    lib/swift/iphoneos/armv7s/libswiftCore.dylib
    lib/swift/iphoneos/armv7/libswiftCore.dylib
```

gribozavr commented on utils/build-script-impl:
```diff
@@ -922,138 +1121,156 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+                ;;
+        esac
+
+        SWIFT_SDKS+=("${swift_sdk}")
```
> OK I've tested that we can build on OSX and it configures for all available SDKs but only builds OSX.

I tested your branch with:

```
$ ./utils/build-script -R --distcc --build-stdlib-deployment-targets=macosx-x86_64
```

And I see build-script-impl configuring Swift like this:

```
-DSWIFT_SDKS:STRING=IOS;IOS_SIMULATOR;OSX;TVOS;TVOS_SIMULATOR;WATCHOS;WATCHOS_SIMULATOR
```

I think you need to wrap everything starting from this line until the end of the loop in `if [ "${is_in_build_list}" ]]`.  This would get us the desired behavior.

karwa replied:
Yes it will configure for all SDKs because that's what is given in `stdlib-deployment-targets` (see `utils/swift-build-supports/targets.py`). However, it will only build for `macosx-x86_64`.

As I've implemented it now, `stdlib-deployment-targets` controls which targets get configured and `build-stdlib-deployment-targets` controls which get built (default is `all`, meaning all configured, although this is in conjunction with the various `skip-ios`..etc flags as we have them today -- it's an additional filter).

This means you can add the `--ios` flag, for example, and if you have any in your build list (or you didn't restrict the build list) it will build the iOS slices without needing to be reconfigured.

gribozavr replied:
> Yes it will configure for all SDKs because that's what is given in `stdlib-deployment-targets` (see `utils/swift-build-support/targets.py`). However, it will only build for `macosx-x86_64`.

I see.  In that case, the change from `swift-sdks=OSX` to `build-stdlib-deployment-targets=macosx-x86_64` in build-presets.ini is not an equivalent one.  Could you adjust that to be `stdlib-deployment-targets` -- that's the new equivalent if I understand correctly?

karwa replied:
Yes, that's correct.

gribozavr replied:
Thank you, that fixed the issue.

gribozavr commented on utils/build-script-impl:
```diff
@@ -100,7 +100,7 @@ KNOWN_SETTINGS=(
     test-installable-package    ""               "whether to run post-packaging tests on the produced package"
     reconfigure                 ""               "force a CMake configuration run even if CMakeCache.txt already exists"
-    swift-sdks                  ""               "build target binaries only for specified SDKs (semicolon-separated list)"
+    build-stdlib-deployment-targets "all"        "build target binaries only for specified SDKs (semicolon-separated list)"
```
It feels really strange to call it `build-*`, since it is about configuration.  Or did the semantics change?

gribozavr commented:
I'm building a toolchain locally to verify that there are no changes in the installed content.

gribozavr commented:
@karwa Could you resolve the conflicts, please?

karwa commented:
LGTM

gribozavr commented:
@karwa Sorry, the build did not finish because of an error while archiving:

```
-- Create Info.plist --
-- Removing: <srcroot>/swift/swift-nightly-install//Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/Info.plist
File Doesn't Exist, Will Create: <srcroot>/swift/swift-nightly-install/Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/Info.plist
tar: Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain: Cannot stat: No such file or directory
```

I used the command printed by `./utils/build-toolchain local.swift` and added `--skip-test-osx --skip-test-lldb --skip-test-swift` to it.

karwa commented:
There are some strange things in those presets (and the toolchain building code in general), which reference some kind of install-directory of `/Library/Developer/Toolchains/<toolchain name>`. This path doesn't get used anywhere - it isn't in the Info.plist or encoded anywhere in the resulting toolchain, it's not even a writeable location unless you're root, and it's not necessary to put toolchains there in any case - they can quite happily go in a user-scoped `~/Library/Developer/Toolchains` directory.

gribozavr commented:
@karwa This command:

```
tar -c -z -f "${package_for_host}" "${TOOLCHAIN_PREFIX/#\/}"
```
used to archive the `.../swift-nightly-install/Library/Developer/Toolchains/` directory.  Now `$TOOLCHAIN_PREFIX` does not seem to be a full path, so `tar` fails.

karwa commented:
OK, it expands to:
`--toolchain-prefix=/Library/Developer/Toolchains/swift-LOCAL-2016-05-26-a.xctoolchain --install-prefix=/usr`

Which should be correct.

rintaro commented on utils/build-script-impl:
```diff
@@ -705,14 +731,89 @@ if [[ "${CHECK_ARGS_ONLY}" ]]; then
+    SKIP_BUILD_IOS_SIMULATOR=1
+    SKIP_TEST_IOS_HOST=1
+    SKIP_TEST_IOS_SIMULATOR=1
+fi
```
Accidentally add backed? #2562 migrated this part to Python.

karwa replied:
Yep. This patch is getting difficult to keep merged. Fixed.

gribozavr commented:
And another issue is that now build-script-impl creates a doubly-nested directory structure in the installation directory:

```
swift-nightly-install/Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/usr/bin/swift
```

karwa commented:
Okay, took a long time to build the preset to test this fix, but we can work with paths for toolchain prefixes as before 👍

gribozavr commented on utils/build-script-impl:
```diff
@@ -2190,17 +2429,17 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
                         set_lldb_build_mode
                         pushd ${LLDB_SOURCE_DIR}
-                        xcodebuild -target toolchain -configuration ${LLDB_BUILD_MODE} install ${lldb_xcodebuild_options[@]} DSTROOT="${INSTALL_DESTDIR}" LLDB_TOOLCHAIN_PREFIX="${TOOLCHAIN_PREFIX}"
+                        xcodebuild -target toolchain -configuration ${LLDB_BUILD_MODE} install ${lldb_xcodebuild_options[@]} DSTROOT="$(get_host_install_destdir_nonprefixed ${host})" LLDB_TOOLCHAIN_PREFIX="/${TOOLCHAIN_PREFIX}"
```
Currently `$TOOLCHAIN_PREFIX` always starts with a slash (see `utils/build-toolchain`), do we need to add an extra one?

karwa replied:
Changed it so `host_install_destdir(_nonprefixed)` always ends in a slash - it is, after all, a directory.

gribozavr commented:
@karwa Thank you for resolving conflicts, I'm building a toolchain again.

gribozavr commented:
@karwa Thank you, we are really close.  The issue I just got is that `swiftc` is located in a funny path:

```
swift-nightly-install/Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/Library/Developer/Toolchains/swift-LOCAL-2016-05-25-a.xctoolchain/usr/bin/swiftc
```

I see you just pushed an update, so I'll rebuild again.

karwa commented:
Update was just a comment fix

karwa commented:
I can't reproduce. I built the same preset and it produced a perfectly fine tree:
[good_tree.txt](https://github.com/apple/swift/files/283581/good_tree.txt)

gribozavr commented:
@karwa Indeed, I can't reproduce either -- sorry!  Probably I did something wrong last time.  I'll try building a Linux toolchain tomorrow, and if that goes well, we should be ready to merge.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@karwa Could you take a look at the buildbot error?

```
error: --install-destdir is required
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/utils/build-script: command terminated with a non-zero exit status 1, aborting
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/utils/build-script: using preset 'buildbot_all_platforms,tools=RA,stdlib=RA', which expands to 

/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/utils/build-script --test --validation-test --ios --tvos --watchos --build-subdir=buildbot_incremental --release --assertions --llbuild --swiftpm -- --reconfigure --verbose-build --build-ninja --build-swift-stdlib-unittest-extra --compiler-vendor=apple --skip-test-ios-host --skip-test-tvos-host --skip-test-watchos-host
```

karwa commented:
@rintaro Failure on Linux is because the python parts of the build script try, like they do for Darwin targets, to configure for all target SDKs. So when you try to build `linux-x86_64`, it will configure for both the host and for `android-armv7`, in the same way it configures for `macosx-x86_64` and for `iphoneos-arm64 iphonesimulator-x86_64 ...etc`. But if you don't have the Android NDK, we're not able to configure the Android target in CMake and that's considered an error.

I'm not sure it really makes sense to automatically configure Android for all Linux targets the same way we do for iOS. The iOS and OSX SDKs come as one package with Xcode, so everybody building for a Mac can always configure for iOS as well. Most people on `linux-x86_64` platforms won't be building for Android, so we shouldn't configure for it unless you explicitly ask for it IMO.

rintaro commented:
@karwa I think, the original intent of `--stdlib_deployment_targets` is
just a list of stdlib targets which the native host *can* build.
This part was in `build-script-impl`, then migrated in #2187.
And that change introduced `--stdlib_deployment_targets`.

I think, this argument is not intended to be used by users, but just for
sharing data between `build-script` and `build-script-impl`

gribozavr commented:
> Most people on `linux-x86_64` platforms won't be building for Android, so we shouldn't configure for it unless you explicitly ask for it IMO.

I think that's the right way to go.  Cross-compiling to iOS from OS X is more of an exception than the rule, but it is a critically important usecase on OS X.

gribozavr commented:
> I think, this argument is not intended to be used by users, but just for sharing data between `build-script` and `build-script-impl`

Currently (when it is called `--swift-sdks`) it is intended to be used by users who have broken SDKs that they don't even want to configure.  This is a rare usecase.  Usually it is put in a preset that some automated system uses.

rintaro commented:
Currently (Prior to this PR):

`--swift-sdks` is to limit auto detected SDKs to **configure**.
SDK auto detection is done in CMakeLists.txt

`--skip-build-{target_variant}` is to limit auto detected stdlib targets to **build**.
The target list auto detection is done in `build-script` and passed to `build-script-impl` using `--stdlib_deployment_targets`

Right?

rintaro commented:
I saw https://github.com/apple/swift/pull/2497/files#r64635763 this comment right now.
You modified the semantics of  `--stdlib-deployment-targets` from "build" to "configure".
If so, I think, you should also modify Python code
from `stdlib-deployment-target` to `build-stdlib-deployment-target` .

gribozavr commented:
@karwa I was looking into merging this patch and realized that something that would be really nice (and make it easier to merge the patch) is to add support to build-script to temporarily continue accepting `swift-sdks` and translate that to the new arguments.  We have internal presets that rely on `swift-sdks`.  I would really appreciate if you could make this change.

karwa commented:
OK, the update I just pushed fixes the OSX buildbot (no longer errors when no install-destdir given).

I'll work on the `swift-sdks` issue

gribozavr commented on utils/build-script-impl:
```diff
@@ -2124,9 +2348,29 @@ for deployment_target in "${STDLIB_DEPLOYMENT_TARGETS[@]}"; do
+    # If INSTALL_DESTDIR not given, we can't install anything.
+
+    if [[ -z "${INSTALL_DESTDIR}" ]] ; then
+        echo "--install-destdir was not given. Not installing any products."
```
Sorry, this is an unfortunate behavior change.  Previously if the user requested installing something (via `--install-*` flags) but did not provide a directory, we produced an error.  Now we would just skip the installation, which is probably not what they wanted.  I'm afraid this would hide errors or make it harder that necessary to debug issues.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@karwa Just wanted to give you a heads up about a possibly conflicting PR in flight: https://github.com/apple/swift/pull/2739/files

karwa commented:
> --swift-sdks is to limit auto detected SDKs to configure.
> SDK auto detection is done in CMakeLists.txt

Calling it 'auto detection' is being a bit generous. If you don't explicitly set a list of SDKs to configure, it'll try to configure everything based on what other parameters you've set (e.g. if you've given it an Android NDK path, it'll configure the Android SDK). So really, it isn't auto detection at all - it just detects to configure for Android based on the presence of an `ANDROID_NDK_PATH` parameter rather than the `SWIFT_SDKS` parameter. It doesn't save you the need to be explicit about configuring Android, it just saves you needing the know the correct parameter for which targets get configured and it doesn't fail if you forgot to set some other required parameter.

Right now we're always explicit about which SDKs to configure for. You're right that this breaks auto detection from CMake. Eventually auto detection should probably be moved in to the build script, anyway:

- SDK names aren't architecture-specific, so eventually `build-script-impl` is going to have to start configuring swift using `stdlib-deployment-targets` such as `linux-armv7` instead of just `LINUX`, or `iphoneos-arm64` instead of just `IOS`.
- Once we do that, we can remove the hardcoded architectures from CMake. They're all in the build script anyway, so we _can_ be explicit (actually somebody forgot `iphoneos-armv7s` in `targets.py`, but otherwise they're all there).
- In fact, if CMake doesn't have hardcoded architectures for its SDKs, it will require us to be explicit about the targets to configure.

gribozavr commented:
> Eventually auto detection should probably be moved in to the build script, anyway

I agree.

karwa commented:
OK, so:
- Install issue has been fixed. Now errors only if trying to install and --install-destdir not given, as before.
- Updated description of `stdlib-deployment-targets`
Python changes:
- I don't know what you'd call it... symbol-ising? symbolicating? the deployment target string literals. Brought `iphoneos-armv7s` back and added `allTargets` helpers so we don't forget any in the future.
- `swift-sdks` gets migrated for presets (tested by reverting the change to `build-presets.ini`, it gets expanded correctly)
- `stdlib_deployment_targets()` is now `default_stdlib_deployment_targets()`. For all platforms other than OSX, it just returns the host's deployment target.
- `DeploymentTarget.Android.armv7` added on request instead of being part of the default list for linux

gribozavr commented on CMakeLists.txt:
```diff
@@ -470,7 +470,7 @@ function(is_sdk_requested name result_var_name)
     if("${SWIFT_SDKS}" STREQUAL "")
-      set("${result_var_name}" "TRUE" PARENT_SCOPE)
+      set("${result_var_name}" "FALSE" PARENT_SCOPE)
     else()
```
Can we just drop the `if` part now?  If the list is empty, the item is not in the list, and the `else` branch will return false.

gribozavr commented on utils/swift_build_support/swift_build_support/targets.py:
```diff
@@ -12,98 +12,127 @@
-            return ['cygwin-x86_64']
-
-    return None
+class StdlibDeploymentTarget:
```
Could you switch to use new-style Python classes?

`class StdlibDeploymentTarget(object):`

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@karwa I'm getting this error now:

```
ninja: error: unknown target 'swift-benchmark-iphoneos-armv7s', did you mean 'swift-benchmark-iphoneos-armv7'?
```

Seems to be caused by this line in `benchmark/CMakeLists.txt`:

```
set(iphoneos_arch "arm64" "armv7")
```

karwa commented:
OK, so either armv7s was intentionally left out of the stdlib-deployment-targets (strange if true, probably deserves a comment to explain), or the benchmarks need to be updated.

@rintaro ?

gribozavr commented:
I think there is a possibility it might have been intentionally left out of benchmarks because armv7 is not dramatically different from armv7s.

karwa commented on utils/build-script-impl:
```diff
@@ -926,138 +1067,163 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+                # Never build iOS armv7s benchmarks.
+                if [[ "${stdlib_deployment_target}" == "iphoneos-armv7s" ]]; then
+                    build_benchmark_this_target=
+                fi
```
Hmm, so we weren't explicitly building for `armv7s` before, but our CMake configuration makes all iOS architectures depend on each other so it just ends up getting built anyway. I wouldn't really like to add this back; it basically relies on a bug/accidental behaviour.

Unfortunately, the cleanest way to skip benchmarks for one particular stdlib target but build others in the same SDK right now is to hardcode it in the build script.

gribozavr replied:
I think this is a clean enough solution.

gribozavr commented:
@swift-ci Please test

karwa commented:
_of course_ swift-build-support has tests which must be updated, too...

Oh well. Let this be the final fix. 😣

gribozavr commented:
@karwa Did you accidentally drop this change?

```
-                # Never build iOS armv7s benchmarks.
-                if [[ "${stdlib_deployment_target}" == "iphoneos-armv7s" ]]; then
-                    build_benchmark_this_target=
-                fi
```

karwa commented:
@gribozavr Yes; too eager to get this merged.

gribozavr commented:
@karwa Sorry, just noticed another thing -- I'm not seeing where the `build_and_test_installable_package` function ends.  Did we lose the close brace?

rintaro commented on utils/build-script-impl:
```diff
@@ -1689,6 +1899,7 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
                             LLDB_PATH_TO_CLANG_BUILD="${llvm_build_dir}"
                             LLDB_PATH_TO_SWIFT_BUILD="${swift_build_dir}"
                             LLDB_PATH_TO_CMARK_BUILD="${cmark_build_dir}"
+                            LLDB_PATH_TO_SWIFTC:PATH="$(build_directory ${LOCAL_HOST} swift)/bin/swiftc"
```
Do we need ":PATH" for `xcodebuild`?

karwa replied:
No, we can get rid of it. This parameter doesn't even get used by `xcodebuild`; the REPL gets compiled by a python script.

karwa commented:
Yes, we lost a `}` with the last merge. Sorry about that.

gribozavr commented:
@karwa I looked into this further and I think we lost more than just the brace.  I see these diffs (compared to yesterday's checkout):

```
           if [[ "${DARWIN_TOOLCHAIN_INSTALLER_PACKAGE}" ]] ; then
             echo "-- Create Installer --"
             "${SWIFT_SOURCE_DIR}/utils/toolchain-installer" "${install_destdir}" "${DARWIN_TOOLCHAIN_BUNDLE_IDENTIFIER}" \
             "${DARWIN_TOOLCHAIN_INSTALLER_CERT}" "${DARWIN_TOOLCHAIN_INSTALLER_PACKAGE}" "${DARWIN_TOOLCHAIN_INSTALL_LOCATION}" \
             "${DARWIN_TOOLCHAIN_VERSION}" "${SWIFT_SOURCE_DIR}/utils/darwin-installer-scripts"
           fi

-          # host_install_destdir contains the toolchain prefix.
-          # We want to create the package in host_install_destdir_nonprefixed.
-          (cd "$(get_host_install_destdir_nonprefixed {host})" &&
-            tar -c -z -f "${package_for_host}" "${TOOLCHAIN_PREFIX/#\/}")
-        else
-            # tar on OSX doesn't support --owner/--group.
-            if [[ "$(uname -s)" == "Darwin" ]] ; then
-                (cd "${install_destdir}" &&
-                        tar -c -z -f "${package_for_host}" "${INSTALL_PREFIX/#\/}")
-            else
-                (cd "${install_destdir}" &&
-                        tar -c -z -f "${package_for_host}" --owner=0 --group=0 "${INSTALL_PREFIX/#\/}")
-            fi
+      (cd "${INSTALL_DESTDIR}" &&
+        tar -c -z -f "${INSTALLABLE_PACKAGE}" "${TOOLCHAIN_PREFIX/#\/}")
+    else
+      (cd "${INSTALL_DESTDIR}" &&
+        tar -c -z -f "${INSTALLABLE_PACKAGE}" --owner=0 --group=0 "${INSTALL_PREFIX/#\/}")
+    fi
+    if [[ "${TEST_INSTALLABLE_PACKAGE}" ]] ; then
+        PKG_TESTS_SOURCE_DIR="${WORKSPACE}/swift-integration-tests"
+        PKG_TESTS_SANDBOX_PARENT="$(build_directory swift_package_sandbox none)"
+        PKG_TESTS_TEMPS="${PKG_TESTS_SANDBOX_PARENT}"/"tests"
+
+        if [[ "$(uname -s)" == "Darwin" ]] ; then
+            PKG_TESTS_SANDBOX="${PKG_TESTS_SANDBOX_PARENT}"/"${TOOLCHAIN_PREFIX}"
+        else # Linux
+            PKG_TESTS_SANDBOX="${PKG_TESTS_SANDBOX_PARENT}"
         fi
-        if [[ "${TEST_INSTALLABLE_PACKAGE}" ]] ; then
-            PKG_TESTS_SOURCE_DIR="${WORKSPACE}/swift-integration-tests"
-            PKG_TESTS_SANDBOX_PARENT="$(build_directory swift_package_sandbox_${host} none)"
-
-            if [[ "${host}" == "macosx-"* ]] ; then
-                PKG_TESTS_SANDBOX="${PKG_TESTS_SANDBOX_PARENT}"/"${TOOLCHAIN_PREFIX}"
-            else # Linux
-                PKG_TESTS_SANDBOX="${PKG_TESTS_SANDBOX_PARENT}"
-            fi
             LIT_EXECUTABLE_PATH="${LLVM_SOURCE_DIR}/utils/lit/lit.py"
             FILECHECK_EXECUTABLE_PATH="$(build_directory_bin ${LOCAL_HOST} llvm)/FileCheck"
             echo "-- Test Installable Package --"
             set -x
             rm -rf "${PKG_TESTS_SANDBOX_PARENT}"
             mkdir -p "${PKG_TESTS_SANDBOX}"
             pushd "${PKG_TESTS_SANDBOX_PARENT}"
             tar xzf "${package_for_host}"
             popd

-            (cd "${PKG_TESTS_SOURCE_DIR}" &&
-                    python "${LIT_EXECUTABLE_PATH}" . -sv --param package-path="${PKG_TESTS_SANDBOX}" --param filecheck="${FILECHECK_EXECUTABLE_PATH}")
-            { set +x; } 2>/dev/null
-        fi
+        (cd "${PKG_TESTS_SOURCE_DIR}" &&
+                python "${LIT_EXECUTABLE_PATH}" . -sv --param package-path="${PKG_TESTS_SANDBOX}" --param filecheck="${FILECHECK_EXECUTABLE_PATH}" --param test-exec-root="${PKG_TESTS_TEMPS}")
+        { set +x; } 2>/dev/null
     fi
 }
```

I don't think this is intended.  The indentation is off, and the `tar` command is back to using `TOOLCHAIN_PREFIX`.  Could you double-check?

karwa commented:
Oh gosh, I'm incredibly sorry; it's a difficult patch to keep merged but that's still a big oversight.

I've re-rebased and double-checked.

gribozavr commented:
@karwa Thank you, I'm building toolchains to verify.

karwa commented:
I don't know, everything is working for me except the LLDB install location, which still forgets to put a path separator before the toolchain prefix. My next idea is that it's a problem with Xcode, as per [Xcode's documentation](https://developer.apple.com/library/mac/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW23):

 ```
INSTALL_DIR
Description:
Directory path. Identifies the directory in the developer’s filesystem into which the installed product is placed.
Effectors:
DSTROOT (Installation Build Products Location), INSTALL_PATH (Installation Directory).
Value:
$DSTROOT/INSTALL_PATH
Example value:
/tmp/MyProduct.dst/Users/genica/Library/Bundles
```

Meanwhile, I'm seeing:
```
export DSTROOT=/Spring/Projects/3rdParty/OpenSource/Apple/cross-compile/out/destdir
...
export INSTALL_PATH=swift3.xctoolchain/System/Library/PrivateFrameworks
...
export INSTALL_DIR=/Spring/Projects/3rdParty/OpenSource/Apple/cross-compile/out/destdirswift3.xctoolchain/System/Library/PrivateFrameworks
```

with parameters:
```
--install-destdir="${cc_files}/out/destdir"
--toolchain_prefix="swift3.xctoolchain"
```

I'm pretty sure I should go back to ignoring the `LLDB_TOOLCHAIN_PREFIX` parameter. While I was changing the build presets, it occurred to me that we never even used to use the `TOOLCHAIN_PREFIX` parameter - we just slapped it on the front of `INSTALL_PREFIX` and pretended there was no `TOOLCHAIN_PREFIX`. We do a similar thing now in the build script - appending on to `INSTALL_DESTDIR` instead of prefixing the prefix. I think that means we should be able to safely ignore the parameter.

Indeed, that's what fixed it before.

gribozavr commented:
@karwa I'm seeing this issue now (tested the PR before the LLDB change):

```
-- Test Installable Package --
[...]
Testing Time: 19.18s
  Expected Passes    : 11
  Unsupported Tests  : 8
--- Creating symbols package ---
-- Package file: <src>/swift/swift-LOCAL-2016-05-27-a-osx-symbols.tar.gz --
+ pushd <src>/swift/swift-nightly-symroot
OSError: [Errno 2] No such file or directory: '<src>swift/swift-nightly-symroot'
```

gribozavr commented:
> While I was changing the build presets, it occurred to me that we never even used to use the `TOOLCHAIN_PREFIX` parameter

I think we were.  From https://ci.swift.org/job/oss-swift-package-osx/1704/consoleText:

```
Build settings from command line:
    LLDB_TOOLCHAIN_PREFIX = /Applications/Xcode.app/Contents/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-05-26-a.xctoolchain
```

But that's a total lie, the toolchain does not end up being installed there.  Given that, I think something about LLDB should have changed since the time when this option was introduced.

/CC @scallanan: what do you think?

karwa commented on utils/build-script-impl:
```diff
@@ -2267,146 +2531,178 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
+        # Strip executables, shared libraries and static libraries in
+        # `host_install_destdir`.
+        find "${host_install_destdir}" \
+          -perm -0111 -or -name "*.a" -type f -print | \
```
The previous version of this command included brackets. That doesn't seem to work (OSX 10.11.4); `find` returns too many files, including scripts and executables which are codesigned and can't be stripped.

Luckily we don't seem to need to brackets.

gribozavr replied:
Why wasn't this an issue before?

gribozavr commented:
@karwa Rebuilt from scratch again, still hitting

```
--- Creating symbols package ---
-- Package file: <src>/swift/swift-LOCAL-2016-05-27-a-osx-symbols.tar.gz --
+ pushd <src>/swift/swift-nightly-symroot
OSError: [Errno 2] No such file or directory: '<src>swift/swift-nightly-symroot'
```

karwa commented:
Turns out that `--toolchain-prefix` isn't a parameter anymore. `build-script` now invents it by extracting the first component of the `--install-prefix`. That will probably explain the LLDB issues, and it's why it can't find the symroot package afterwards.

In that case it's probably best that I revert the `--install-prefix` change and add another parameter: `cross-compile-host-install-prefix`. We're not going to be able to share `--install-prefix` with other platforms.

gribozavr commented:
@karwa Your proposed plan sounds reasonable.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please smoke test

karwa commented:
@gribozavr Okay, changes to `TOOLCHAIN_PREFIX` reverted and it builds a correctly laid-out toolchain on OSX.

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented:
@karwa Thank you, I'm building a toolchain locally to double-check.

karwa commented:
@gribozavr Hadn't pushed all commits

gribozavr commented:
Fetched again.

gribozavr commented:
@karwa Thanks!  The install tree looks a little funny, Swift does not seem to be installed into the toolchain:

```
swift-nightly-install/Library/Developer/Toolchains/swift-LOCAL-2016-05-27-a.xctoolchain/usr/bin/lldb # LLDB is OK
swift-nightly-install/usr/bin/swift # Swift is not
```

karwa commented:
swift will be inside the toolchain:
```
swift-nightly-install/Library/Developer/Toolchains/swift-LOCAL-2016-05-27-a.xctoolchain/usr/bin/swift
```

gribozavr commented:
@karwa Sorry, I don't understand you...  I built a toolchain and the file list I posted is what I got.  Are you saying that you are working on a fix?

karwa commented:
No, I'm saying that swift is installed where it should be - inside the toolchain directory. The path you posted was in the install_destdir itself.

This is the tree I got from `build_toolchain`:
[toolchain.txt](https://github.com/apple/swift/files/289189/toolchain.txt)

gribozavr commented:
Looks like this is the issue:

```shell
DESTDIR="${host_install_destdir}" "${CMAKE}" --build "${build_dir}" -- ${INSTALL_TARGETS}
```

Should it use `${host_install_destdir}${host_install_prefix}` instead?

gribozavr commented:
> The path you posted was in the install_destdir itself.

Well, yes, that's the issue I'm pointing out.  The tree I posted is what I got from build-toolchain, too.

gribozavr commented:
@swift-ci Please test

karwa commented:
No I'm still pretty sure this is right - the install prefix wasn't in the `DSTDIR` variable before these changes, either. Swift gets this information from the `CMAKE_INSTALL_PREFIX` variable.

If you have an extra copy of swift, maybe it's left over from another build?

gribozavr commented:
Let's see if it reproduces in CI.

gribozavr commented:
> No I'm still pretty sure this is right - the install prefix wasn't in the `DSTDIR` variable before these changes, either. Swift gets this information from the `CMAKE_INSTALL_PREFIX` variable.

But when we call the CMake action to install, we override the directory with `DSTDIR` environment variable.

> If you have an extra copy of swift, maybe it's left over from another build?

It is not an extra copy, it is the only copy.  (And I'm always doing from-scratch builds, removing the `../build` directory and `swift-nightly-*` directories.)

gribozavr commented:
I will also build another toolchain without this PR to check which commands were run before.

karwa commented on utils/build-script-impl:
```diff
@@ -1546,32 +1767,32 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_COMPILE_TOOLS_DEPLOYMENT_TARG
                     "${cmake_options[@]}"
-                    -DCMAKE_INSTALL_PREFIX:PATH="${INSTALL_PREFIX}"
-                    -DLLVM_CONFIG:PATH="$(build_directory_bin ${deployment_target} llvm)/llvm-config"
+                    -DCMAKE_INSTALL_PREFIX:PATH="$(get_host_install_prefix ${host})"
```
This is where we set the install prefix for swift. It's in CMakeCache.txt. Should be:
```
CMAKE_INSTALL_PREFIX:PATH=/Library/Developer/Toolchains/swift-LOCAL-2016-05-29-a.xctoolchain/usr
```

gribozavr replied:
Yes, this is exactly right, and it is what I'm seeing.  So the issue is either somewhere else, or it is my local misconfiguration.  See my other comment -- I'm running another build.

gribozavr commented:
> > No I'm still pretty sure this is right - the install prefix wasn't in the `DSTDIR` variable before these changes, either. Swift gets this information from the `CMAKE_INSTALL_PREFIX` variable.

> But when we call the CMake action to install, we override the directory with `DESTDIR` environment variable.

I finally see what you meant, sorry, I misunderstood you.  I'm building again now.  I can confirm that `CMAKE_INSTALL_PREFIX` for LLVM and Swift are correct, now waiting for the installation step.

gribozavr commented:
I just built the toolchains again on OS X and Linux, and everything matches up exactly!  So it was a misconfiguration on my side, sorry for wasting your time, @karwa.

karwa commented:
@gribozavr Thank you so much for all your help and for reviewing it. Also to @rintaro and everybody else.

jrose-apple commented:
I can't tell for sure, but it looks like this commit introduced Linux triples of the form "armv7-linux-gnueabihf" instead of "armv7-unknown-linux-gnueabihf", causing [SR-3019](https://bugs.swift.org/browse/SR-3019). Was this intentional?

karwa commented:
@jrose nope. I picked that to be as close as possible to the GCC triple arm-linux-gnueabihf (but I don't think anybody uses armv6/7 on GCC). I wasn't aware that triples in LLVM *needed* a fourth component.

jrose-apple commented:
They need three components, but the second component has to be a vendor. Are you sure the GCC triple is arm-linux-gnueabihf and not arm-unknown-linux-gnueabihf?

jrose-apple commented:
Okay, I could have just checked for that, indeed it is.

karwa commented:
Yeah sorry, you're right that the second component should be "unknown" (rustc  apparently uses that), but GCC seems to use an incorrect triple: http://packages.ubuntu.com/precise/devel/gcc-arm-linux-gnueabihf

I've been having some issues with LLDB reloading images because of mismatching architectures. I'll investigate if this could be the reason for it...

jrose-apple commented:
I doubt it is, but we should pull in a fix anyway.
